capsules/src/virtual_ninedof.rs

```diff
@@ -0,0 +1,141 @@
+//! This allows a 9DOF sensor to be used by multiple apps.
+//!
+//! The data from the driver is not virtualized. This just gives apps
+//! exclusive access to the driver until a callback occurs.
+
+use core::cell::Cell;
+use kernel::{AppId, Callback, Container, Driver};
+use kernel::hil;
+use kernel::returncode::ReturnCode;
+
+pub struct App {
+    callback: Option<Callback>,
+    pending_command: bool,
+    command: usize,
+    arg1: usize,
+}
+
+impl Default for App {
+    fn default() -> App {
+        App {
+            callback: None,
+            pending_command: false,
+            command: 0,
+            arg1: 0,
+        }
+    }
+}
+
+pub struct VirtualNineDof<'a> {
+    driver: &'a hil::ninedof::NineDof,
+    apps: Container<App>,
+    current_app: Cell<Option<AppId>>,
+}
+
+impl<'a> VirtualNineDof<'a> {
+    pub fn new(driver: &'a hil::ninedof::NineDof, container: Container<App>) -> VirtualNineDof<'a> {
+        VirtualNineDof {
+            driver: driver,
+            apps: container,
+            current_app: Cell::new(None),
+        }
+    }
+
+    fn enqueue_command(&self, command_num: usize, arg1: usize, appid: AppId) -> ReturnCode {
+        self.apps
+            .enter(appid, |app, _| {
+                // Check so see if we are doing something. If not,
+                // go ahead and do this command. If so, this is queued
+                // and will be run when the pending command completes.
+                if self.current_app.get().is_none() {
+                    self.current_app.set(Some(appid));
+                    self.call_driver(command_num, arg1)
+                } else {
+                    if app.pending_command == true {
+                        ReturnCode::ENOMEM
+                    } else {
+                        app.pending_command = true;
+                        app.command = command_num;
+                        app.arg1 = arg1;
+                        ReturnCode::SUCCESS
+                    }
+                }
+            })
+            .unwrap_or(ReturnCode::FAIL)
+    }
+
+    fn call_driver(&self, command_num: usize, _: usize) -> ReturnCode {
+        match command_num {
+            1 => self.driver.read_accelerometer(),
```

Comment:
Given this driver is now specific to ninedof, it seems like this could be hardened by using a specific enum instead of propogating the `command_num` all the way through (i.e. in `Driver#command` enqueue and enum constructor instead of `1` and `100`). May also make sense to inline this function in the `NineDofClient` if it becomes even smaller.